File: app/src/main/java/com/artt/rotationcontrolv2/RotationService.java
package com.artt.rotationcontrolv2;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.IBinder;
import android.provider.Settings;
import android.provider.Settings.System;
import android.support.annotation.Nullable;
import android.view.Surface;
import android.view.WindowManager;
import android.widget.RemoteViews;
import android.widget.Toast;

public class RotationService extends Service implements AccelerometerRotationObserver.Callback {

    public static boolean isStarted = false;

    //private static final String TAG = "RotationControlV2";
    private static final String INTENT_CLICKED_BUTTON_ID = "BUTTON_ID";
    private static final String INTENT_ACTION = "ACTION";
    private static final String NOTIFICATION_CHANNEL = "Rotation Control";
    private static final String NOTIFICATION_CHANNEL_ID = "rotation_service_notification_channel_id";

    private static final int NOTIFICATION_ID = 999;
    private static final int ACTION_CLICK = 2;
    private static final int ORIENTATION_PORTRAIT = Surface.ROTATION_0;
    private static final int ORIENTATION_LANDSCAPE = Surface.ROTATION_90;
    private static final int ORIENTATION_PORTRAIT_REVERSE = Surface.ROTATION_180;
    private static final int ORIENTATION_LANDSCAPE_REVERSE = Surface.ROTATION_270;
    private static final int ORIENTATION_AUTO = 4;
    private final int[] IDS_BUTTON = new int[]{R.id.btn_auto, R.id.btn_portrait, R.id.btn_portrait_reverse,
            R.id.btn_landscape, R.id.btn_landscape_reverse};

    private int mScreenOrientation;
    private AccelerometerRotationObserver mAccelerometerRotationObserver;

    @Override
    public void update() {
        if (mScreenOrientation != ORIENTATION_AUTO && PermissionManager.CanWriteSettings(getApplicationContext()))
            System.putInt(getContentResolver(), System.ACCELEROMETER_ROTATION, 0);
    }

    private void setOrientation(int orientation){
        if (!PermissionManager.CanWriteSettings(getApplicationContext())) {
            Toast.makeText(this, R.string.service_write_denied, Toast.LENGTH_LONG).show();
            return;
        }

        if (mScreenOrientation != ORIENTATION_AUTO) {
            System.putInt(getContentResolver(), System.USER_ROTATION, orientation);

            try {
                if (System.getInt(getContentResolver(), System.ACCELEROMETER_ROTATION) == 1)
                    System.putInt(getContentResolver(), System.ACCELEROMETER_ROTATION, 0);
            } catch (Settings.SettingNotFoundException e) {
                e.printStackTrace();
            }
        }
        else
            System.putInt(getContentResolver(), System.ACCELEROMETER_ROTATION, 1);
    }

    @Override
    public void onCreate() {
        super.onCreate();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL_ID,
                    NOTIFICATION_CHANNEL, NotificationManager.IMPORTANCE_HIGH);
            if (manager != null)
                manager.createNotificationChannel(channel);
        }

        WindowManager manager = (WindowManager) getSystemService(Context.WINDOW_SERVICE);

        if (manager != null)
            mScreenOrientation = manager.getDefaultDisplay().getRotation();
        else {
            stopSelf();
            return;
        }

        mAccelerometerRotationObserver = new AccelerometerRotationObserver(this);
        getContentResolver().registerContentObserver(System.getUriFor(System.ACCELEROMETER_ROTATION),
                false, mAccelerometerRotationObserver);

        createNotification(mScreenOrientation);

        isStarted = true;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        getContentResolver().unregisterContentObserver(mAccelerometerRotationObserver);

        isStarted = false;
    }

    private void onClick(int id){
        switch (id){
            case R.id.btn_auto:{
                mScreenOrientation = ORIENTATION_AUTO;
                break;
            }
            case R.id.btn_portrait:{
                mScreenOrientation = ORIENTATION_PORTRAIT;
                break;
            }
            case R.id.btn_portrait_reverse:{
                mScreenOrientation = ORIENTATION_PORTRAIT_REVERSE;
                break;
            }
            case R.id.btn_landscape: {
                mScreenOrientation = ORIENTATION_LANDSCAPE;
                break;
            }
            case R.id.btn_landscape_reverse:{
                mScreenOrientation = ORIENTATION_LANDSCAPE_REVERSE;
                break;
            }
        }
        createNotification(mScreenOrientation);
        setOrientation(mScreenOrientation);
    }

    private void createNotification(int orientation) {
        RemoteViews notificationContent = new RemoteViews(getPackageName(), R.layout.notification);
        PendingIntent activity = PendingIntent.getActivity(this, 0,
                new Intent(this, SettingsActivity.class), Intent.FILL_IN_COMPONENT);

        Intent intent = new Intent(this, RotationService.class);
        intent.putExtra(INTENT_ACTION, ACTION_CLICK);
        for (int length = IDS_BUTTON.length, i = 0; i < length; ++i) {
            int id = IDS_BUTTON[i];
            intent.putExtra(INTENT_CLICKED_BUTTON_ID, id);
            notificationContent.setOnClickPendingIntent(id, PendingIntent.getService(this, id,
                    intent, PendingIntent.FLAG_UPDATE_CURRENT));
            notificationContent.setInt(id, "setBackgroundResource", android.R.color.transparent);
            notificationContent.setInt(id, "setColorFilter", getResources().getColor(R.color.btn_active));
        }

        int smallIconId = -1;
        int buttonId = -1;
        switch (orientation){
            case ORIENTATION_AUTO:{
                buttonId = R.id.btn_auto;
                smallIconId = R.drawable.ic_stat_notify_auto;
                break;
            }
            case ORIENTATION_PORTRAIT: {
                buttonId = R.id.btn_portrait;
                smallIconId = R.drawable.ic_stat_notify_portrait;
                break;
            }
            case ORIENTATION_PORTRAIT_REVERSE:{
                buttonId = R.id.btn_portrait_reverse;
                smallIconId = R.drawable.ic_stat_notify_portrait_reverse;
                break;
            }
            case ORIENTATION_LANDSCAPE: {
                buttonId = R.id.btn_landscape;
                smallIconId = R.drawable.btn_landscape;
                break;
            }
            case ORIENTATION_LANDSCAPE_REVERSE: {
                buttonId = R.id.btn_landscape_reverse;
                smallIconId = R.drawable.btn_landscape_reverse;
                break;
            }
        }

        notificationContent.setInt(buttonId, "setBackgroundResource", R.drawable.btnback_active);
        notificationContent.setInt(buttonId, "setColorFilter", android.R.color.transparent);

        Notification notification;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notification = new Notification.Builder(this, NOTIFICATION_CHANNEL_ID)
                    .setCustomContentView(notificationContent)
                    .setContentIntent(activity)
                    .setSmallIcon(smallIconId)
                    .build();
            notification.flags = Notification.FLAG_ONLY_ALERT_ONCE;
        }
        else
            notification = new Notification.Builder(this)
                    .setContent(notificationContent)
                    .setContentIntent(activity)
                    .setSmallIcon(smallIconId)
                    .setPriority(Notification.PRIORITY_MAX)
                    .build();


        showNotification(notification);
    }

    private void showNotification(Notification notification){
        startForeground(NOTIFICATION_ID, notification);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            int action = intent.getIntExtra(INTENT_ACTION, 0);
            switch (action){
                case ACTION_CLICK:
                    onClick(intent.getIntExtra(INTENT_CLICKED_BUTTON_ID, 0));
            }
        }

        return START_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/arttttt/rotationcontrolv3/model/services/RotationService.kt
package com.arttttt.rotationcontrolv3.model.services

import android.app.*
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings.System
import com.arttttt.rotationcontrolv3.model.services.base.ServiceHelper
import android.view.Surface
import com.arttttt.rotationcontrolv3.R
import android.widget.RemoteViews
import android.app.PendingIntent
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.content.res.Configuration
import android.database.ContentObserver
import android.support.v4.app.NotificationCompat
import android.support.v4.content.ContextCompat
import android.widget.Toast
import com.arttttt.rotationcontrolv3.model.permissions.base.PermissionsChecker
import com.arttttt.rotationcontrolv3.utils.OsUtils
import org.koin.standalone.StandAloneContext


class RotationService: Service() {
    override fun onBind(intent: Intent?) = null

    companion object: ServiceHelper {
        private val koinContext = StandAloneContext.getKoin().koinContext
        private val mStarted = MutableLiveData<Boolean>()
        private val permissionsChecker: PermissionsChecker = koinContext.get()

        init {
            mStarted.value = false
        }

        override fun isStarted(): LiveData<Boolean> = mStarted

        override fun startService(context: Context) {
            if (!permissionsChecker.canWriteSettings(context)) {
                Toast.makeText(context, R.string.can_not_write_settings_toast, Toast.LENGTH_LONG).show()

                return
            }

            Intent(context, RotationService::class.java).let {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(it)
                } else {
                    context.startService(it)
                }
            }
        }

        override fun stopService(context: Context) {
            Intent(context, RotationService::class.java).let {
                context.stopService(it)
            }
        }
    }

    private val intentAction = "ACTION"
    private val intentActionClick = 2
    private val intentClickedButtonId = "BUTTON_ID"

    private val notificationChannel = "Rotation Control"
    private val notificationChannelId = "rotation_service_notification_channel_id"
    private val notificationId = 999

    private val orientationPortrait = Surface.ROTATION_0
    private val orientationLandscape = Surface.ROTATION_90
    private val orientationPortraitReverse = Surface.ROTATION_180
    private val orientationLandscapeReverse = Surface.ROTATION_270
    private val orientationAuto = 4

    private val accelerometerObserver = object: ContentObserver(null) {
        override fun onChange(selfChange: Boolean) {
            super.onChange(selfChange)
            checkCurrentMode()
        }
    }

    private var mScreenOrientation = orientationAuto

    private val buttonsData = arrayOf(
        Pair(R.id.btn_auto, R.drawable.ic_portrait_auto),
        Pair(R.id.btn_portrait, R.drawable.ic_portrait),
        Pair(R.id.btn_portrait_reverse, R.drawable.ic_portrait_reverse),
        Pair(R.id.btn_landscape, R.drawable.ic_landscape),
        Pair(R.id.btn_landscape_reverse, R.drawable.ic_landscape_reverse))

    override fun onCreate() {
        super.onCreate()

        mStarted.value = true

        contentResolver.registerContentObserver(System.getUriFor(System.ACCELEROMETER_ROTATION),
            false,
            accelerometerObserver)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val channel = NotificationChannel(
                notificationChannelId,
                notificationChannel,
                NotificationManager.IMPORTANCE_HIGH
            )
            manager.createNotificationChannel(channel)
        }

        mScreenOrientation = convertOrientation(resources.configuration.orientation)
        setOrientation()

        showNotification(createNotification())
    }

    override fun onDestroy() {
        super.onDestroy()

        mStarted.value = false
        contentResolver.unregisterContentObserver(accelerometerObserver)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        intent?.let {
            val action = it.getIntExtra(intentAction, 0)
            when(action) {
                intentActionClick -> { handleNotificationClick(it.getIntExtra(intentClickedButtonId, 0)) }
            }
        }
        return Service.START_STICKY
    }

    private fun checkCurrentMode() {
        if (!permissionsChecker.canWriteSettings(this) || mScreenOrientation == orientationAuto)
            return

        System.putInt(contentResolver, System.ACCELEROMETER_ROTATION, 0)
    }

    private fun convertOrientation(origOrientation: Int): Int {
        return if (origOrientation == Configuration.ORIENTATION_PORTRAIT)
            orientationPortrait
        else
            orientationLandscape
    }

    private fun createNotification(): Notification {
        val notificationContent = RemoteViews(packageName, R.layout.notification)

        val activeButtonId = when (mScreenOrientation) {
            orientationPortrait -> R.id.btn_portrait
            orientationPortraitReverse -> R.id.btn_portrait_reverse
            orientationLandscape -> R.id.btn_landscape
            orientationLandscapeReverse -> R.id.btn_landscape_reverse
            else ->  R.id.btn_auto
        }

        for (buttonData in buttonsData) {
            val intent = Intent(this, RotationService::class.java).apply {
                putExtra(intentAction, intentActionClick)
                putExtra(intentClickedButtonId, buttonData.first)
            }
            notificationContent.setOnClickPendingIntent(buttonData.first, PendingIntent.getService(this,
                buttonData.first,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT))

            if (buttonData.first == activeButtonId) {
                notificationContent.setInt(buttonData.first,
                    "setBackgroundResource",
                    R.drawable.active_button_background)

                val colorRes = if (OsUtils.isMiui()) R.color.colorActiveButtonTintMiui else R.color.colorActiveButtonTint

                notificationContent.setInt(buttonData.first,
                    "setColorFilter",
                    ContextCompat.getColor(this, colorRes))
            } else {
                notificationContent.setInt(buttonData.first,
                    "setBackgroundResource",
                    android.R.color.transparent)

                val colorRes = if (OsUtils.isMiui()) R.color.colorInactiveButtonTintMiui else R.color.colorInactiveButtonTint

                notificationContent.setInt(buttonData.first,
                    "setColorFilter",
                    ContextCompat.getColor(this, colorRes))
            }
        }

        return NotificationCompat.Builder(this, notificationChannelId)
            .setCustomContentView(notificationContent)
            .setSmallIcon(R.drawable.ic_rotate)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .apply {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
                    setStyle(NotificationCompat.DecoratedCustomViewStyle())
            }
            .build()
            .apply {
                flags = NotificationCompat.FLAG_ONLY_ALERT_ONCE
            }
    }

    private fun handleNotificationClick(buttonId: Int) {
        if (!permissionsChecker.canWriteSettings(this)) {
            Toast.makeText(this, R.string.can_not_write_settings_toast, Toast.LENGTH_LONG).show()

            stopSelf()
            return
        }

        mScreenOrientation =  when (buttonId) {
            R.id.btn_portrait -> orientationPortrait
            R.id.btn_portrait_reverse -> orientationPortraitReverse
            R.id.btn_landscape -> orientationLandscape
            R.id.btn_landscape_reverse -> orientationLandscapeReverse
            else -> orientationAuto
        }
        setOrientation()
        showNotification(createNotification())
    }

    private fun showNotification(notification: Notification) = startForeground(notificationId, notification)

    private fun setOrientation() {
        if (!permissionsChecker.canWriteSettings(this))
            return

        if (mScreenOrientation == orientationAuto)
            System.putInt(contentResolver, System.ACCELEROMETER_ROTATION, 1)
        else {
            checkCurrentMode()
            System.putInt(contentResolver, System.USER_ROTATION, mScreenOrientation)
        }
    }
}